docs/proposals/node-local-services.md

```diff
@@ -0,0 +1,260 @@
+    ServiceAffinityClientIP ServiceAffinity = "ClientIP"
+
+    // ServiceAffinityNone - no session affinity.
+    ServiceAffinityNone ServiceAffinity = "None"
+
+    <b>// ServiceAffinityRequireNodeLocal - only talk to a pod on the same
+    // node, typically managed by a Daemonset.
+    ServiceAffinityRequireNodeLocal ServiceAffinity = "RequireNodeLocal"</b>
+)
+
+type ServiceSpec struct {
+    Ports []ServicePort
+
+    // If not specified, the associated Endpoints object is not automatically managed
+    Selector map[string]string
+
+    // "", a real IP, or "None".  If not specified, this is default allocated.  If "None", this Service is not load-balanced
+    ClusterIP string
+
+    // Type determines how the service will be exposed.  Valid options: ClusterIP, NodePort, LoadBalancer
+    Type ServiceType
+
+    // Only applies if clusterIP != "None"
+    ExternalIPs []string
+
+    <b>// Optional, used to maintain session affinity.
+    // Supports "ClientIP", "RequireNodeLocal" and "None". The second makes
+    // most sense with `Type: ClusterIP`.</b>
+    SessionAffinity ServiceAffinity
+
+    // Only applies to type=LoadBalancer
+    LoadBalancerIP string
+    LoadBalancerSourceRanges []string
+</pre>
+
+### `kube-proxy` changes
+
+The most fundamental modification necessary for the proxy is becoming aware of
+the node's IP address(es). Until now, it has behaved identically across all
+nodes in the cluster: e.g., in iptables mode, it creates the same set of rules
+everywhere, so node identity has never been required. Discovering the right
+address might be tricky, especially if starting from a node name and/or
+considering differences between cloud providers — or between those and bare
+metal environment. Because the proxy is currently not linked against cloud
+providers and in order to avoid doing so, a compromise would be to match the
+list of endpoints in a service against the addresses of all the network
+interfaces on the node. Should there be concerns about the cost of looking up
+all the interfaces on a regular basis, a simple one-entry cache (an `<address,
+interface>` pair, with the full search as a fallback) might obviate a lot of the
+extra overhead.
+
+#### iptables kernel mode
+
+In the existing implementation, `kube-proxy` keeps track of all N endpoints for
+a service. On each node, it creates `2*N` iptables rules to pick a random
+endpoint when opening a new connection: one rule for the probabilistic selection
+and another one for the actual `DNAT` packet rewriting. For node-local services,
+it will require only one rule or even zero, if it keeps track of the actual pod
+liveness. The end result is less overhead over a conventional service, before
+even taking into account network latency.
+
+For use cases like `kube2iam`, which requires intercepting traffic meant for the
+special `169.254.169.254` metadata IP address, perhaps `ClusterIP` should be
+allowed to live outside of the service IP range.
```

Comment:
I've added this to the implementation plan and the fact that this would require an explicit flag somewhere in the control plane, due to its security implications.